src/gc.c

```diff
@@ -3393,6 +3401,8 @@ static int _jl_gc_collect(jl_ptls_t ptls, jl_gc_collection_t collection)
         gc_num.last_incremental_sweep = gc_end_time;
     }
 
+    jl_atomic_store_relaxed(&gc_heap_stats.heap_size, jl_atomic_load_relaxed(&gc_heap_stats.heap_size) - freed_in_runtime);
```

Comment:
I'm inclined to subtract `freed_in_runtime` in the computation of `before_free_heap_size`? (And possibly rename the `before_free_heap_size ` variable)

In the GMP case, the Nth GC cycle runs, Nth finalizer cycle runs and frees up some memory.

It seems that the freeing of memory through finalizers should be reflected in the heap size of the beginning of the (N+1)th GC cycle.